libraries/stdlib/samples/test/samples/misc/preconditions.kt

```diff
@@ -27,6 +27,22 @@ class Preconditions {
         assertPrints(getIndices(3), "[1, 2, 3]")
     }
 
+    @Sample
+    fun failRequireNotNullWithLazyMessage() {
+
+        fun execWithParams(params: Map<String, String?>) {
+            val required = requireNotNull(params["required"]) { "Required key must be non-null" } // returns a non-null value
```

Comment:
Though, if you have a bit more time, could you add another variant of `requireNotNull` usage that utilizes its contract allowing to smartcast its parameter. Something like:
```
val requiredParam: String? = params["required"]
requireNotNull(requiredParam) { "Required value must be non-null" }
// now requiredParam is smartcast to String
```

I think it could be done with another local function.

Reply:
OK, I work on it soon.